substrate/frame/assets/src/tests.rs

```diff
@@ -1345,9 +1345,9 @@ fn freezer_should_work() {
 		// freeze 50 of it.
 		set_frozen_balance(0, 1, 50);
 
-		assert_ok!(Assets::transfer(RuntimeOrigin::signed(1), 0, 2, 20));
-		// cannot transfer another 21 away as this would take the non-frozen balance (30) to below
-		// the minimum balance (10).
+		assert_ok!(Assets::transfer(RuntimeOrigin::signed(1), 0, 2, 30));
```

Comment:
This is to reflect the change in decreaseable balance, to match the [balance components' model](https://paritytech.github.io/polkadot-sdk/master/frame_support/traits/tokens/fungible/index.html#visualising-balance-components-together-).

That way, assertions for `transfer` calls below don't need to change.

Reply:
I see we have a bug, `ed + frozen_balance` is considered as non transferable (with keep alive), but is should be `ed.max(frozen_balance)`.

Reply:
Here's a more detailed explanation of what's happening in this test.

### Explanation

Consider what happens in this case (I'll try to follow because there are multiple changes 😅):

For every step, `ed` is `10` (line [1341](https://github.com/paritytech/polkadot-sdk/pull/4530/files/aab6498e500041573120ba38fb6732cbfeefec9f#diff-156573d6bcc9270c4fde34e94bb4b71f8c702f6702d1132302810464f8c8ebf8L1341))

**Step 1 (lines [1342-1346](https://github.com/paritytech/polkadot-sdk/pull/4530/files/aab6498e500041573120ba38fb6732cbfeefec9f#diff-156573d6bcc9270c4fde34e94bb4b71f8c702f6702d1132302810464f8c8ebf8R1342-R1346)):**

```
|____________total____________| < 100
|____frozen____| < 50
|__ed__| < 10
|__untouchable_|___spendable__| < untouchable: 50 / spendable (a.k.a. reducible): 50
```

**Step 2 (lines [1348-1354](https://github.com/paritytech/polkadot-sdk/pull/4530/files/aab6498e500041573120ba38fb6732cbfeefec9f#diff-156573d6bcc9270c4fde34e94bb4b71f8c702f6702d1132302810464f8c8ebf8R1348-R1354)):**

```
|____________total____________| < 70
|____frozen____| < 50
|__ed__| < 10
|__untouchable_|___spendable__| < untouchable: 50 / spendable (a.k.a. reducible): 20
```

> Note that from here, further `assert_noop`s are preserved.

**Step 3 (lines [1357-1364](https://github.com/paritytech/polkadot-sdk/pull/4530/files/aab6498e500041573120ba38fb6732cbfeefec9f#diff-156573d6bcc9270c4fde34e94bb4b71f8c702f6702d1132302810464f8c8ebf8R1357-R1364)):** No changes

**Step 4 (lines [1366-1367](https://github.com/paritytech/polkadot-sdk/pull/4530/files/aab6498e500041573120ba38fb6732cbfeefec9f#diff-156573d6bcc9270c4fde34e94bb4b71f8c702f6702d1132302810464f8c8ebf8L1366-L1367)):**

```
|____________total____________| < 70
|____frozen___| < 49
|__ed__| < 10
|_untouchable_|___spendable___| < untouchable: 49 / spendable (a.k.a. reducible): 21
```

**Step 5 (lines [1368-1369](https://github.com/paritytech/polkadot-sdk/pull/4530/files/aab6498e500041573120ba38fb6732cbfeefec9f#diff-156573d6bcc9270c4fde34e94bb4b71f8c702f6702d1132302810464f8c8ebf8L1368-L1369)):**

```
|____total____| < 49
|____frozen___| < 49
|__ed__| < 10
|_untouchable_| < untouchable: 49 / spendable (a.k.a. reducible): 0
```

**Step 6 (lines [1371-1372](https://github.com/paritytech/polkadot-sdk/pull/4530/files/aab6498e500041573120ba38fb6732cbfeefec9f#diff-156573d6bcc9270c4fde34e94bb4b71f8c702f6702d1132302810464f8c8ebf8L1371-L1372)):**

```
|____total____| < 49
|__ed__| < 10
|__ut__|__sp__| < untouchable: 10 / spendable (a.k.a. reducible): 39
```

**Step 7 (lines [1373-1374](https://github.com/paritytech/polkadot-sdk/pull/4530/files/aab6498e500041573120ba38fb6732cbfeefec9f#diff-156573d6bcc9270c4fde34e94bb4b71f8c702f6702d1132302810464f8c8ebf8L1373-L1374)):** _Died_.

> Note: `Assets::transfer` [sets the `keep_alive` flag](https://github.com/pandres95/polkadot-sdk/blob/548eee10766f6fc24dea9d475799c90881a2176c/substrate/frame/assets/src/lib.rs#L970) to `false`. (Maybe that's something that should be changed, since we already have `transfer_all`?)

### Conclusion

In this test, you can see there wasn't a calculation of `untouchable` of `ed + frozen`, but it's always `max(ed, frozen)`.

This can be seen here: [`substrate/frame/assets/functions.rs:247-248`](https://github.com/pandres95/polkadot-sdk/blob/548eee10766f6fc24dea9d475799c90881a2176c/substrate/frame/assets/src/functions.rs#L247-L248).

Reply:
I do understand the test. But If we change the expected output and the test that is not related to the PR and is still green, something did changed in the logic. In a diff of `reducible_balance` I can see that it was calculating `required` by summing ed and frozen balance.

Reply:
Yes, that all makes sense. Commenting to make sure this is a deliberate, thoughtful change.

Reply:
Done (on [f1c0983](https://github.com/pandres95/polkadot-sdk/blob/f1c0983b768ad671ade34ca5726b93c1947415e7/substrate/frame/assets/src/tests.rs#L1348)!